### PR TITLE
virttest.virt_vm: add wrapper function for get_address

### PR DESCRIPTION
**Pull request by xutian**
Param 'flexible_nic_index' should also works to get IPv6 address of nic

Signed-off-by: Xu Tian <xutian@redhat.com>

### CODE REVIEW

File: virttest/virt_vm.py [hunk 1]
```diff
@@ -644,6 +644,28 @@ def get_mac_address(self, nic_index=0):
...
     def get_address(self, index=0, ip_version="ipv4"):
         """
+        Wrapper for self._get_address. if 'flexible_nic_index' is 'yes',
+        will traverses from first nic to first element toward the end
+        util get a reachable IP address;
+        """
+        flexible_nic_index = ("yes" == self.params.get("flexible_nic_index", "no"))
+        macs, nics_index = self.virtnet.mac_list(), [index]
+        if flexible_nic_index:
+            nics_index += [i for i in xrange(len(macs)) if i != index]
```
Comment (luckyh) [inline virttest/virt_vm.py]: Looks like you need a `set()`.

File: virttest/virt_vm.py [hunk 2]
```diff
@@ -644,6 +644,28 @@ def get_mac_address(self, nic_index=0):
 
     def get_address(self, index=0, ip_version="ipv4"):
         """
+        Wrapper for self._get_address. if 'flexible_nic_index' is 'yes',
+        will traverses from first nic to first element toward the end
+        util get a reachable IP address;
+        """
```
Comment (yanan-fu) [inline virttest/virt_vm.py]: params and  the type ?

File: virttest/virt_vm.py [hunk 3]
```diff
@@ -644,6 +644,28 @@ def get_mac_address(self, nic_index=0):
...
+        flexible_nic_index = ("yes" == self.params.get("flexible_nic_index", "no"))
+        macs, nics_index = self.virtnet.mac_list(), [index]
+        if flexible_nic_index:
+            nics_index += [i for i in xrange(len(macs)) if i != index]
+        flexible_nic_index &= bool(len(macs) > 1)
```
Comment (luckyh) [inline virttest/virt_vm.py]: Better to combine this with line 651 together to reduce loop overhead.

File: virttest/virt_vm.py [hunk 4]
```diff
@@ -644,6 +644,28 @@ def get_mac_address(self, nic_index=0):
...
+        for index in nics_index:
+            try:
+                return self._get_address(index, ip_version)
+            except VMIPAddressMissingError:
```
Comment (luckyh) [inline virttest/virt_vm.py]: What about `VMAddressVerificationError` and `VMMACAddressMissingError` two type of errors .

File: virttest/virt_vm.py [hunk 5]
```diff
@@ -644,6 +644,28 @@ def get_mac_address(self, nic_index=0):
...
+                if not flexible_nic_index:
+                    raise
+                else:
+                    continue
+        else:
+            raise
```
Comment (luckyh) [inline virttest/virt_vm.py]: Can we raise an `IndexError` or some else?

File: virttest/virt_vm.py (hunk 1 above)
Comment (xutian) [inline virttest/virt_vm.py]: No,  index should be try first, so use list here.

File: virttest/virt_vm.py (hunk 3 above)
Comment (xutian) [inline virttest/virt_vm.py]: agree, with fix in next version

File: virttest/virt_vm.py (hunk 5 above)
Comment (xutian) [inline virttest/virt_vm.py]: It's better to raise real exception here.

File: virttest/virt_vm.py (hunk 4 above)
Comment (xutian) [inline virttest/virt_vm.py]: agree, will fix in next version.

**Comment by xutian** [comment]
@luckyh  updated, would you like to help review it, thanks.

File: virttest/virt_vm.py (hunk 1 above)
Comment (luckyh) [inline virttest/virt_vm.py]: Alright.

File: virttest/virt_vm.py (hunk 5 above)
Comment (luckyh) [inline virttest/virt_vm.py]: But the "real" exception is a little misleading. If we get here, it means no reachable IP addrs have been found for all the NICs, not just for the last one.

File: virttest/virt_vm.py [hunk 6]
```diff
@@ -644,6 +644,28 @@ def get_mac_address(self, nic_index=0):
...
+        """
+        flexible_nic_index = ("yes" == self.params.get("flexible_nic_index", "no"))
+        macs, nics_index = self.virtnet.mac_list(), [index]
+        flexible_nic_index &= bool(len(macs) > 1)
+        if flexible_nic_index:
+            nics_index += [i for i in xrange(len(macs)) if i != index]
+        for index in nics_index:
+            try:
+                return self._get_address(index, ip_version)
+            except (VMIPAddressMissingError, VMMACAddressMissingError):
```
Comment (luckyh) [inline virttest/virt_vm.py]: I think, when `flexible_nic_index` is `True`, `VMAddressVerificationError` could be ignored like it used to be, until it found a valid address.

File: virttest/virt_vm.py (hunk 5 above)
Comment (xutian) [inline virttest/virt_vm.py]: Sounds reasonable, let's raise VMAddressError("VM '%s' no reachable %s address found" % (self.name, ip_version) ).
what's your opinion?

Comment (luckyh) [inline virttest/virt_vm.py]: Looks good.

File: virttest/virt_vm.py (hunk 4 above)
Comment (xutian) [inline virttest/virt_vm.py]: right, it should catch VMAddressError. fixed in new version. thanks

File: virttest/virt_vm.py (hunk 6 above)
Comment (xutian) [inline virttest/virt_vm.py]: right, here should ignore VMAddressError, has updated in new version.

**Comment by xutian** [comment]
@luckyh  any other comments for this version?

**Comment by CongLi** [comment]
ID: 1442685

**Comment by CongLi** [comment]
Acked-by: Cong Li <coli@redhat.com>

@xutian 
Could you please check @yanan-fu comment to add doc string for param in get_address() ?

Thanks.

**Comment by xutian** [comment]
@CongLi and @yanan-fu  as doc string said it's wrapper for self._get_address, I think it's needless to add duplicate docs in get_address function.

**Comment by xutian** [comment]
thanks @CongLi and @luckyh

**Pull request merged**